x/examples/outline-cli/main.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build linux
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/signal"
+	"sync"
+
+	"golang.org/x/sys/unix"
+)
+
+const OUTLINE_TUN_NAME = "outline233"
+const OUTLINE_TUN_IP = "10.233.233.1"
+const OUTLINE_TUN_MTU = 1500 // todo: we can read this from netlink
+const OUTLINE_GW_SUBNET = "10.233.233.2/32"
+const OUTLINE_GW_IP = "10.233.233.2"
+const OUTLINE_ROUTING_PRIORITY = 23333
+const OUTLINE_ROUTING_TABLE = 233
+const OUTLINE_DNS_SERVER = "9.9.9.9"
+
+// ./app -transport "ss://..."
+func main() {
+	fmt.Println("OutlineVPN CLI (experimental)")
+
+	transportFlag := flag.String("transport", "", "Transport config")
+	flag.Parse()
+
+	// this WaitGroup must Wait() after tun is closed
+	trafficCopyWg := &sync.WaitGroup{}
+	defer trafficCopyWg.Wait()
+
+	tun, err := NewTunDevice(OUTLINE_TUN_NAME, OUTLINE_TUN_IP)
+	if err != nil {
+		log.Printf("[error] failed to create tun device: %v\n", err)
+		return
+	}
+	defer tun.Close()
+
+	// disable IPv6 before resolving Shadowsocks server IP
+	prevIPv6, err := enableIPv6(false)
+	if err != nil {
+		log.Printf("[error] failed to disable IPv6: %v\n", err)
+		return
+	}
+	defer enableIPv6(prevIPv6)
+
+	ss, err := NewOutlineDevice(*transportFlag)
+	if err != nil {
+		log.Printf("[error] failed to create Outline device: %v", err)
+		return
+	}
+	defer ss.Close()
+
+	ss.Refresh()
+
+	// Copy the traffic from tun device to OutlineDevice bidirectionally
+	trafficCopyWg.Add(2)
+	go func() {
+		defer trafficCopyWg.Done()
+		written, err := io.Copy(ss, tun)
+		log.Printf("[info] tun -> OutlineDevice stopped: %v %v\n", written, err)
+	}()
+	go func() {
+		defer trafficCopyWg.Done()
+		written, err := io.Copy(tun, ss)
+		log.Printf("[info] OutlineDevice -> tun stopped: %v %v\n", written, err)
+	}()
+
+	if err := setSystemDNSServer(OUTLINE_DNS_SERVER); err != nil {
+		log.Printf("[error] failed to configure system DNS: %v", err)
+		return
+	}
+	defer restoreSystemDNSServer()
+
+	if err := startRouting(ss.GetServerIP().String(),
+		OUTLINE_TUN_NAME,
```

Comment:
Perhaps it's an improvement for later, but we need to separate what are implementation details of a VPN api, and what are parameters a client should pass to a VPN API.

For instance, a client shouldn't have to worry about tun name, routing table, routing priority, but should be able to tell the IP to use and what gets rounted, as well as the DNS servers to use.

I suggest you look into the Android and iOS APIs for what should be in a VPN API, and create a type for that.

Reply:
Yes, I can see the interface will definitely change over time. For example, Windows `startRouting` will not use these settings. For now since it's not a public library or API, I'll just create a `RoutingConfig` type to wrap around these `string`s so it's more readable.

Reply:
Also the `TUN_IP` and `GW_IP` and `TUN_MTU` should be randomly chosen. The user should not hard code that (I cannot think of any scenarios that a caller must provide a hard-coded `TUN_IP`).